File: adventofcode/2022/day5/part1.py
import os
from collections import defaultdict
import pprint

current_dir = os.path.dirname(os.path.abspath(__file__))


def main():
    input_file_path = os.path.join(current_dir, "input")
    _input = open(input_file_path, "r").read()
    input_lines = _input.split("\n")

    input_file_path = os.path.join(current_dir, "input-ins")
    _input = open(input_file_path, "r").read()
    input_ins_lines = _input.split("\n")

    # Parse input stack
    stacks = defaultdict(list)
    for line in input_lines[::-1]:
        i = 0
        while i < len(line):
            _next = line[i:i + 4]
            if _next.strip() == "":
                pass
            else:
                _next = _next.strip().lstrip("[").rstrip("]")
                if _next:
                    stacks[(i // 4) + 1].append(_next)
            i += 4

    # Parse intructions
    instructions = []
    for line in input_ins_lines:
        line = line.split()
        instructions.append([int(line[1]), int(line[3]), int(line[5])])

    # make the moves
    for moves, _from, _to in instructions:
        for i in range(moves):
            val = stacks[_from].pop()
            stacks[_to].append(val)

    # output
    output = ""
    for i, v in stacks.items():
        output += v[-1]
    print(output)

if __name__ == "__main__":
    main()

File: adventofcode/2022/day5/part1.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	inputData, _ := os.ReadFile("input")
	inputInstructionData, _ := os.ReadFile("input-ins")
	inputs := strings.Split(string(inputData), "\n")
	inputInstructions := strings.Split(string(inputInstructionData), "\n")

	// Parse the input data
	cols := (len(inputs[0]) + 1) / 4
	stacks := make([][]string, cols)
	for j := range inputs {
		line := inputs[len(inputs)-j-1]
		lineArr := strings.Split(line, "")
		i := 0
		for i < len(lineArr) {
			val := lineArr[i+1]
			if val != " " {
				stacks[i/4] = append(stacks[i/4], val)
			}
			i += 4
		}
	}

	// Parse the input instructions
	instructions := [][3]int{}
	for _, instructionLine := range inputInstructions {
		parts := strings.Split(instructionLine, " ")
		moves, _ := strconv.Atoi(parts[1])
		from, _ := strconv.Atoi(parts[3])
		to, _ := strconv.Atoi(parts[5])
		instruction := [3]int{moves, from, to}
		instructions = append(instructions, instruction)
	}

	// Make the moves
	for _, instruction := range instructions {
		moves := instruction[0]
		from := instruction[1]
		to := instruction[2]
		for i := 0; i < moves; i++ {
			_from := stacks[from-1]
			val := _from[len(_from)-1]
			stacks[from-1] = _from[:len(_from)-1]
			stacks[to-1] = append(stacks[to-1], val)
		}
	}

	// Output
	output := ""
	for _, s := range stacks {
		output += s[len(s)-1]
	}
	fmt.Println(output)

}
